src/Core/Content/Media/Upload/MediaUploadService.php

```diff
@@ -0,0 +1,223 @@
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Criteria;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Filter\EqualsFilter;
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Core\Framework\Util\Hasher;
+use Shopware\Core\Framework\Uuid\Uuid;
+use Symfony\Component\HttpFoundation\File\UploadedFile;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @final
+ */
+#[Package('discovery')]
+readonly class MediaUploadService
+{
+    /**
+     * @internal
+     */
+    public function __construct(
+        private EntityRepository $mediaRepository,
+        private FileFetcher $fileFetcher,
+        private FileSaver $fileSaver,
+        private EventDispatcherInterface $eventDispatcher,
+        private HttpClientInterface $httpClient,
+    ) {
+    }
+
+    /**
+     * Upload a new media file from a local path
+     */
+    public function uploadFromLocalPath(string $filePath, Context $context, MediaUploadParameters $params = new MediaUploadParameters()): string
+    {
+        $size = filesize($filePath);
+
+        if ($size === false) {
+            throw MediaException::fileNotFound($filePath);
+        }
+
+        $media = new MediaFile(
+            $filePath,
+            mime_content_type($filePath) ?: '',
+            pathinfo($filePath, \PATHINFO_EXTENSION),
+            $size,
+            Hasher::hashFile($filePath, 'md5'),
+        );
+
+        return $this->upload($media, $context, $params);
+    }
+
+    /**
+     * Upload a new media file provided as form-data in the Request object
+     */
+    public function uploadFromRequest(Request $request, Context $context, MediaUploadParameters $params = new MediaUploadParameters()): string
+    {
+        /** @var UploadedFile|null $file */
+        $file = $request->files->get('file');
+
+        if ($file === null) {
+            throw MediaException::fileNotProvided();
+        }
+
+        $params->fillDefaultFileName($file->getClientOriginalName());
+
+        $media = new MediaFile(
+            $file->getRealPath(),
+            (string) $file->getMimeType(),
+            $file->getClientOriginalExtension(),
+            $file->getSize(),
+            Hasher::hashFile($file->getRealPath(), 'md5'),
+        );
+
+        return $this->upload($media, $context, $params);
+    }
+
+    /**
+     * Download the given media file from the URL and upload it as own file
+     */
+    public function uploadFromURL(string $url, Context $context, MediaUploadParameters $params = new MediaUploadParameters()): string
+    {
+        $tempFile = tempnam(sys_get_temp_dir(), '');
+
+        if (!$tempFile) {
+            throw MediaException::cannotCreateTempFile();
+        }
+
+        $params->fillDefaultFileName(basename($url));
+
+        try {
+            $media = $this->fileFetcher->fetchFileFromURL(new Request([], ['url' => $url]), $tempFile);
```

Comment:
This seems hacky, as a new Request instance is created and passed just to satisfy the method signature. Wouldn’t it be cleaner to have a separate method that directly accepts a string URL instead?